agent/agent.go

```diff
@@ -218,6 +213,7 @@ func (a *Agent) run(ctx context.Context) {
 			if a.err == nil {
 				a.err = ctx.Err()
 			}
+			session.close()
```

Comment:
I always get confused about where it's safe to call these close functions and where it's not safe. Is this one of the places where we can do it without worrying about a double close?